torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,346 @@
+  HANDLE_TH_ERRORS
+  PyObject *_inputs, *_outputs, *_streams;
+  int root, op;
+
+  if (!PyArg_ParseTuple(args, "OOOii", &_inputs, &_outputs, &_streams, &root, &op)) {
+    THPUtils_invalidArguments(args, NULL, "nccl_reduce", 1,
+			      "(sequence[Tensor] inputs, sequence[Tensor]"
+			      " outputs, sequence[torch.cuda.Stream or None], int root, int op");
+    return NULL;
+  }
+
+  std::vector<at::Tensor> inputs = THPUtils_PySequence_to_TensorList(_inputs);
+  std::vector<at::Tensor> outputs = THPUtils_PySequence_to_TensorList(_outputs);
+  std::vector<THCStream*> streams = THPUtils_PySequence_to_THCStreamList(_streams);
+
+  THPUtils_assert(inputs.size() == streams.size(), "number of streams is not equal to number of inputs");
+  
+  // we can safely release GIL after this line, no python API used
+  AutoNoGIL no_gil;
+  _check_inputs(inputs, outputs, 1, 1);
+  size_t len = inputs.size();
+
+  ncclDataType_t data_type = _get_data_type(inputs[0].type().ID());
+
+  int64_t count = inputs[0].numel();
+  std::lock_guard<std::mutex> lock(*(THCCachingAllocator_getCudaFreeMutex()));
+  ncclComm_t *comm = _get_communicator(inputs);
+  AutoGPU gpu_guard;
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupStart());
+#endif
+  for (size_t i = 0; i < len; i++) {
+    int device = inputs[i].get_device();
+    gpu_guard.setDevice(device);
+    auto stream = (streams[i] == NULL) ? NULL : streams[i]->stream;
+    CHECK(ncclReduce(inputs[i].data_ptr(), outputs[i].data_ptr(),
+		     count, data_type, (ncclRedOp_t) op, root, comm[i], stream));
+  }
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupEnd());
+#endif
+
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject * THCPModule_nccl_all_reduce(PyObject *self, PyObject *args) {
+  HANDLE_TH_ERRORS
+  PyObject *_inputs, *_outputs;
+  int op;
+
+  if (!PyArg_ParseTuple(args, "OOi", &_inputs, &_outputs, &op)) {
+    THPUtils_invalidArguments(args, NULL, "nccl_all_reduce", 1,
+			      "(sequence[Tensor] inputs, sequence[Tensor]"
+			      " outputs, int op");
+    return NULL;
+  }
+
+  std::vector<at::Tensor> inputs = THPUtils_PySequence_to_TensorList(_inputs);
+  std::vector<at::Tensor> outputs = THPUtils_PySequence_to_TensorList(_outputs);
+
+  // we can safely release GIL after this line, no python API used
+  AutoNoGIL no_gil;
+  _check_inputs(inputs, outputs, 1, 1);
+  size_t len = inputs.size();
+
+  ncclDataType_t data_type = _get_data_type(inputs[0].type().ID());
+
+  int64_t count = inputs[0].numel();
+  std::lock_guard<std::mutex> lock(*(THCCachingAllocator_getCudaFreeMutex()));
+  ncclComm_t *comm = _get_communicator(inputs);
+  AutoGPU gpu_guard;
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupStart());
+#endif
+  for (size_t i = 0; i < len; i++) {
+    int device = inputs[i].get_device();
+    gpu_guard.setDevice(device);
+    CHECK(ncclAllReduce(inputs[i].data_ptr(), outputs[i].data_ptr(),
+			count, data_type, (ncclRedOp_t) op, comm[i], NULL));
```

Comment:
Running on NULL stream, esp when additional streams are created with CudaStreamNonBlocking might lead to synchronization bugs. This is also present in the current nccl python bindings, only reduce handles streams.

Reply:
yea these need to be fixed, but not in the scope of this PR.